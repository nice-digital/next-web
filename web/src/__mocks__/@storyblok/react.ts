export const storyblokInit = (): null => null;

export const apiPlugin = (): null => null;
