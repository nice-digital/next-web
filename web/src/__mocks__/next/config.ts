import nextConfig from "./../../../next.config";

const getConfig = (): unknown => ({
	publicRuntimeConfig: nextConfig.publicRuntimeConfig,
	serverRuntimeConfig: nextConfig.serverRuntimeConfig,
});

export default getConfig;
