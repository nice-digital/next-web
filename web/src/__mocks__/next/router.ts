export const useRouter = jest.fn().mockImplementation(() => {
	throw new Error("Provide a mock of useRouter for testing purposes");
});
