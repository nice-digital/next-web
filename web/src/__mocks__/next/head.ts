/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { FC } from "react";
import { renderToString } from "react-dom/server";

interface HeadProps {
	children: Parameters<typeof renderToString>[0];
}

/**
 * Mocking next/head allows us to use it in our tests to assert against meta tags/titles
 * See https://edibleco.de/3dyJuUl
 */
const Head: FC<HeadProps> = ({ children }: HeadProps) => {
	if (children) {
		global.document.head.insertAdjacentHTML(
			"afterbegin",
			renderToString(children) || ""
		);
	}
	return null;
};

export default Head;
