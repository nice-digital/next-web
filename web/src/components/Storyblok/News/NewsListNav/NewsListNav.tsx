import { useRouter } from "next/router";
import React, { useState, useEffect, useCallback, useMemo } from "react";

import {
	HorizontalNav,
	HorizontalNavLink,
} from "@nice-digital/nds-horizontal-nav";

import { NoScrollLink } from "@/components/Link/Link";

type Destination = {
	url: string;
	title: string;
};

export const NewsListNav = (): React.ReactElement => {
	const router = useRouter();
	const [activeLink, setActiveLink] = useState<Destination>();

	const destinations = useMemo(
		() => [
			{ url: "/news", title: "News" },
			{ url: "/news/articles", title: "News articles" },
			{ url: "/news/in-depth", title: "In-depth" },
			{ url: "/news/blogs", title: "Blogs" },
			{ url: "/news/podcasts", title: "Podcasts" },
		],
		[]
	);

	const getActiveLink = useCallback(
		(pathname: string): Destination | undefined => {
			const baseURL = pathname.split("?")[0]; // Extract base URL without query string
			return destinations.find((link) => baseURL === link.url);
		},
		[destinations]
	);

	useEffect(() => {
		setActiveLink(getActiveLink(router.pathname));
		const handleRouteChange = (url: string) => {
			setActiveLink(getActiveLink(url));
		};

		router.events.on("routeChangeComplete", handleRouteChange);

		return () => {
			router.events.off("routeChangeComplete", handleRouteChange);
		};
	}, [getActiveLink, router]);

	const handleLinkClick = (destination: string) => {
		setActiveLink(getActiveLink(destination));
		router.push({
			pathname: destination,
		});
	};

	return (
		<HorizontalNav aria-label="News section navigation">
			{destinations.map((link, index) => (
				<HorizontalNavLink
					key={index}
					destination={link.url}
					isCurrent={activeLink && activeLink.url === link.url}
					onClick={() => handleLinkClick(link.url)}
					elementType={NoScrollLink}
				>
					{link.title}
				</HorizontalNavLink>
			))}
		</HorizontalNav>
	);
};
