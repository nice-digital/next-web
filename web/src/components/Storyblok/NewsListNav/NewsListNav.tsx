import { useRouter } from "next/router";
import React, { useState, useEffect, useCallback } from "react";

import {
	HorizontalNav,
	HorizontalNavLink,
} from "@nice-digital/nds-horizontal-nav";

import { NoScrollLink } from "@/components/Link/Link";

type Destination = {
	url: string;
	title: string;
};

type DestinationProps = {
	destinations: Destination[];
};

export const NewsListNav = ({
	destinations,
}: DestinationProps): React.ReactElement => {
	const router = useRouter();
	const [activeLink, setActiveLink] = useState<Destination>();

	const getActiveLink = useCallback(
		(pathname: string): Destination | undefined => {
			return destinations.find((link) => pathname === link.url);
		},
		[destinations]
	);

	useEffect(() => {
		setActiveLink(getActiveLink(router.pathname));

		const handleRouteChange = (url: string) => {
			setActiveLink(getActiveLink(url));
		};

		router.events.on("routeChangeComplete", handleRouteChange);

		return () => {
			router.events.off("routeChangeComplete", handleRouteChange);
		};
	}, [getActiveLink, router]);

	const handleLinkClick = (destination: string) => {
		setActiveLink(getActiveLink(destination));
		router.push({
			pathname: destination,
		});
	};

	return (
		<HorizontalNav aria-label="News section navigation">
			{destinations.map((link, index) => (
				<HorizontalNavLink
					key={index}
					destination={link.url}
					isCurrent={activeLink && activeLink.url === link.url}
					onClick={() => handleLinkClick(link.url)}
					elementType={NoScrollLink}
				>
					{link.title}
				</HorizontalNavLink>
			))}
		</HorizontalNav>
	);
};
