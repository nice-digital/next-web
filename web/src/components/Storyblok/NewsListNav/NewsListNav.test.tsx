import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { NextRouter, useRouter } from "next/router";

import { NewsListNav } from "./NewsListNav";

// jest.mock("next/router", () => ({
// 	useRouter: () => ({
// 		route: "/news",
// 		pathname: "/news",
// 		query: {},
// 		asPath: "/news",
// 		events: {
// 			on: jest.fn(),
// 			off: jest.fn(),
// 		},
// 		push: jest.fn(),
// 	}),
// }));

jest.mock("next/router", () => ({
	useRouter: jest.fn(),
}));

describe("NewsListNav", () => {
	const destinations = [
		{ url: "/news", title: "News" },
		{ url: "/blog", title: "Blog" },
		{ url: "/events", title: "Events" },
	];

	let useRouterMock: jest.MockedFunction<typeof useRouter>;
	let router: Partial<NextRouter>;

	beforeEach(() => {
		router = {
			route: "/news",
			pathname: "/news",
			query: {},
			asPath: "/news",
			events: {
				on: jest.fn(),
				off: jest.fn(),
				emit: jest.fn(),
			},
			push: jest.fn(),
		};

		useRouterMock = useRouter as jest.MockedFunction<typeof useRouter>;
		useRouterMock.mockReturnValue(router as unknown as NextRouter);
	});

	it("renders correctly", () => {
		const { asFragment } = render(<NewsListNav destinations={destinations} />);
		expect(asFragment()).toMatchSnapshot();
	});

	it("renders navigation lnks with correct titles", () => {
		render(<NewsListNav destinations={destinations} />);
		destinations.forEach((destination) => {
			expect(screen.getByText(destination.title)).toBeInTheDocument();
		});
	});

	it("renders navigation links with correct href attributes", () => {
		render(<NewsListNav destinations={destinations} />);
		destinations.forEach((destination) => {
			const link = screen.getByText(destination.title);
			expect(link).toBeInTheDocument();
			expect(link.getAttribute("href")).toBe(destination.url);
		});
	});

	it("sets aria-current attribute correctly when link is active", () => {
		render(<NewsListNav destinations={destinations} />);
		const activeLink = screen.getByText("News");
		expect(activeLink).toHaveAttribute("aria-current", "true");
	});

	it("sets does not set aria-current attribute when link is inactive", () => {
		render(<NewsListNav destinations={destinations} />);
		const activeLink = screen.getByText("Blog");
		expect(activeLink).toHaveAttribute("aria-current", "false");
	});

	it("navigates to the correct destination when link is clicked", async () => {
		render(<NewsListNav destinations={destinations} />);
		const blogLink = screen.getByText("Blog") as HTMLAnchorElement;
		await userEvent.click(blogLink);
		await waitFor(() => {
			expect(blogLink.href).toBe("https://next-web-tests.nice.org.uk/blog");
			// eslint-disable-next-line testing-library/no-wait-for-multiple-assertions
			expect(router.push).toHaveBeenCalledWith({
				pathname: "/blog",
			});
		});
	});
});
