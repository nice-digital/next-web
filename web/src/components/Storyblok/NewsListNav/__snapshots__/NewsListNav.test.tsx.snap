// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsListNav renders correctly 1`] = `
<DocumentFragment>
  <nav
    aria-label="News section navigation"
    class="horizontal-nav"
  >
    <ul
      class="horizontal-nav__list"
    >
      <li
        class="horizontal-nav__item"
        data-component="horizontal-nav"
      >
        <a
          aria-current="true"
          class="horizontal-nav__link"
          href="/news"
        >
          News
        </a>
      </li>
      <li
        class="horizontal-nav__item"
        data-component="horizontal-nav"
      >
        <a
          aria-current="false"
          class="horizontal-nav__link"
          href="/blog"
        >
          Blog
        </a>
      </li>
      <li
        class="horizontal-nav__item"
        data-component="horizontal-nav"
      >
        <a
          aria-current="false"
          class="horizontal-nav__link"
          href="/news/articles"
        >
          News articles
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
